src/core/probe/kernel/mod.rs

```diff
@@ -137,30 +168,111 @@ impl Kernel {
     /// kernel.register_hook(hook::DATA)?;
     /// ```
     pub(crate) fn register_hook(&mut self, hook: &'static [u8]) -> Result<()> {
-        if self.hooks.len() == HOOK_MAX {
+        let mut max: usize = 0;
+        for set in self.targeted_probes.iter_mut() {
+            if max < set.hooks.len() {
+                max = set.hooks.len();
+            }
+        }
+        if self.hooks.len() + max == HOOK_MAX {
             bail!("Hook list is already full");
         }
 
         self.hooks.push(hook);
         Ok(())
     }
 
+    /// Request a hook to be attached to a specific target.
+    ///
+    /// ```
+    /// mod hook {
+    ///     include!("bpf/.out/hook.rs");
+    /// }
+    ///
+    /// [...]
+    ///
+    /// kernel.register_hook_to(hook::DATA, ProbeType::Kprobe, "kfree_skb_reason")?;
+    /// ```
+    pub(crate) fn register_hook_to(
+        &mut self,
+        hook: &'static [u8],
+        r#type: ProbeType,
+        target: &str,
+    ) -> Result<()> {
+        let target = target.to_string();
+
+        // First check if the target isn't already registered to the generic
+        // probes list. If so, remove it from there.
+        let target_set = &mut self.probes[r#type as usize];
+        target_set.targets.remove(&target);
+
+        // Now check if we already have a targeted probe for this. If so, append
+        // the new hook to it.
+        for set in self.targeted_probes.iter_mut() {
+            if r#type == set.r#type && set.targets.get(&target).is_some() {
+                if self.hooks.len() + set.hooks.len() == HOOK_MAX {
+                    bail!("Hook list is already full");
+                }
+                set.hooks.push(hook);
+                return Ok(());
+            }
+        }
+
+        // New target, let's build a new probe set.
+        let mut set = ProbeSet::new(
+            r#type,
+            match r#type {
+                ProbeType::Kprobe => Box::new(kprobe::KprobeBuilder::new()),
+                ProbeType::RawTracepoint => Box::new(raw_tracepoint::RawTracepointBuilder::new()),
+                ProbeType::Max => bail!("Invalid probe type"),
+            },
+        );
+
+        let desc = self.inspect_target(&r#type, &target)?;
+        set.targets.insert(target.to_string(), desc);
+
+        if self.hooks.len() == HOOK_MAX {
+            bail!("Hook list is already full");
+        }
+        set.hooks.push(hook);
+
+        self.targeted_probes.push(set);
+        Ok(())
+    }
+
     /// Attach all probes.
     pub(crate) fn attach(&mut self) -> Result<()> {
+        // Take care of target probes first, they are probably higher prio.
+        for set in self.targeted_probes.iter_mut() {
+            let hooks = [set.hooks.clone(), self.hooks.clone()].concat();
```

Comment:
I don't know if it will be the case but I'm thinking that it's more likely that targeted probes make use of information that come from generic probes than the opposite. If so, maybe we should run generic probes first. WDYT?

Reply:
Sorry I didn't respond to this:

> This raises lots of questions:
> 
>     * How do we share info between hooks? Event struct? Custom one? Array of pointers to stack variables w/ an unique id?
> 

I was thinking maybe if a hook may depend on another hook populating some info on the event struct itself.

>     * Is the dependency generic enough to be targeted probes -> generic ones? Or do we need a 1:1 mapping?
> 
>     * If a hook we depend on fail to process, what do to? Do we (and how) propagate the error? The proposal here discard any hook error to let others try gathering info.
> 

No idea. I haven't thought it through too much.

> 
> Do you see not supporting this as a blocking point for your use cases? Is this something we'll have to support, or requesting hooks to be independent would still work?
> 

No, I don't think so.

> IMHO until the above is defined we can ignore the hook calling order. Such information sharing / hook dependency support can come in its own proposal and rework the above part.
> 
> Nit: turns out the comment in the code snippet above is wrong as the hook calling order do not really matter for now...